plonk-hashing/src/poseidon/poseidon.rs

```diff
@@ -0,0 +1,582 @@
+//! optimized poseidon
+
+use crate::poseidon::constants::PoseidonConstants;
+use crate::poseidon::matrix::Matrix;
+use crate::poseidon::mds::SparseMatrix;
+use crate::poseidon::PoseidonError;
+use ark_ec::TEModelParameters;
+use ark_ff::PrimeField;
+use core::{fmt::Debug, marker::PhantomData};
+use derivative::Derivative;
+use plonk_core::constraint_system::{StandardComposer, Variable};
+use plonk_core::prelude as plonk;
+
+// TODO: reduce duplicate code with `poseidon_ref`
+pub trait PoseidonSpec<COM, const WIDTH: usize> {
+    type Field: Debug + Clone;
+    type ParameterField: PrimeField;
+
+    fn output_hash(
+        c: &mut COM,
+        elements: &[Self::Field; WIDTH],
+        constants: &PoseidonConstants<Self::ParameterField>,
```

Comment:
Could we consider adding an index for the output element here? Or maybe some understanding on why the 2nd element is selected and not the 1st (0-indexed element).